src/semanticlayertools/clustering/reports.py

```diff
@@ -0,0 +1,288 @@
+        return corpus_titles
+
+    def find_topics(
+        self, corpus_titles: list, n_topics: int, top_words: int,
+    ):
+        """Calculate topics in corpus.
+
+        Use NMF algorithm to calculate topics in corpus file for `n_topics`
+        topics, returning `top_words` most common words for each topic.
+        Each word has to occure at least twice in the corpus and at most in 95%
+        of all documents.
+
+        :param corpus_titles: The corpus containing the preprocessed texts.
+        :type corpus_titles: `textacy.Corpus`
+        :param n_topics: Number of considered topics
+        :type n_topics: int
+        :param top_words: Number of returned words for each found topic
+        :type top_words: int
+        :returns: List of found topics with top occuring words
+        :rtype: str
+        """
+        vectorizer = textacy.representations.vectorizers.Vectorizer(
+            tf_type="linear",
+            idf_type="smooth",
+            norm="l2",
+            min_df=2,
+            max_df=0.95
+        )
+        tokenized_docs = (
+            (
+                term.lemma_ for term in textacy.extract.terms(doc, ngs=1, ents=True)
+            ) for doc in corpus_titles
+        )
+        doc_term_matrix = vectorizer.fit_transform(tokenized_docs)
+
+        model = textacy.tm.TopicModel("nmf", n_topics)
+        model.fit(doc_term_matrix)
+
+        topics = []
+        for topic_idx, top_terms in model.top_topic_terms(
+            vectorizer.id_to_term, top_n=top_words
+        ):
+            topics.append(
+                "topic " + str(topic_idx) + ": " + "   ".join(top_terms)
+            )
+        outtext = f'\n\n\tTopics in cluster for {n_topics} topics:\n'
+        for topic in topics:
+            outtext += f'\t\t{topic}\n'
+        return outtext
+
+    def fullReport(self, cluster):
+        """Generate full cluster report for one cluster.
+
+        :param cluster: The cluster number to process
+        :type cluster: int or str
+        :raises ValueError: If input cluster data can not be read.
+        :returns: Report text with all gathered informations
+        :rtype: str
+        """
+        starttime = time.time()
+        clusterpath = os.path.join(self.outpath, f'Cluster_{cluster}')
+        clusterfiles = os.listdir(clusterpath)
+        clusterdf = []
+        for x in clusterfiles:
+            try:
+                clusterdf.append(
+                    pd.read_json(os.path.join(clusterpath, x), lines=True)
+                )
+            except ValueError:
+                raise
+        dfCluster = pd.concat(clusterdf, ignore_index=True)
+        basedf = self.clusternodes.query('cluster == @cluster')
+        inputnodes = set(basedf.node.values)
+        notFound = inputnodes.difference(set(dfCluster[self.publicationIDcolumn].values))
+        topAuthors = Counter(
+            [x for y in [x.split(';') for x in dfCluster[self.authorColumnName].fillna('').values] for x in y]
```

Comment:
Double list comprehension always confuses me a lot. I'm never sure which `for` runs first, and what the result is.